cpp/src/parquet/arrow/schema.cc

```diff
@@ -1017,7 +1016,10 @@ Result<bool> ApplyOriginalMetadata(const Field& origin_field, SchemaField* infer
 
       // Restore extension type, if the storage type is the same as inferred
       // from the Parquet type
-      if (ex_type.storage_type()->Equals(*inferred->field->type())) {
+      if (ex_type.storage_type()->Equals(*inferred->field->type()) ||
```

Comment:
So suddently this condition only covers `arrow.json`, while it used to cover all extension types. Is this deliberate?

Reply:
It was, but in light of your comment I'd say it's not a good idea.

Reply:
Hmm, but why are we keeping the logical `||` condition here?

Reply:
Sorry, I only removed the `IsSupportedStorageType` part. The reason being if we have an extension type with `large_utf8` storage the following test bellow will fail, because (if I remember correctly) we take parquet json logical type to mean storage type will be `utf8`. To fix that this proposes to override with inferred type whenever `arrow.json`.

https://github.com/apache/arrow/blob/6d156d1daf09c7f15fb25f95b81a45cce1dc3c71/cpp/src/parquet/arrow/arrow_schema_test.cc#L804-L820

Reply:
That would make it more idiomatic indeed. Changed.